Refactor/cleanup config loading (misc edits)

jdavcs opened the pull request:
Misc. edits. Details/rationale in commit messages.

jdavcs committed:
Replace resolve_path() with os.path.join()

The function was redundant: it checked whether its argument was an
absolute path, and if not, called os.path.join. This is exactly what
os.path.join already does (python 3.x and 2.7)

The only benefit of having a separate function is the ability to add
additinal processing to such calls - so when we need that, we can add
it. Meanwhile, it removes unnecessary function calls + makes the code
slightly less complex.

The function is not called from outside this module (not to be confused
with the resolve_path method of the mixin class), but it's duplicated in
webapps/reports/config.py and webapps/tool_shed/config.py (which are not
addressed in this commit).

jdavcs committed:
Use default value from schema for smtp_ssl

The default value of `smtp_ssl` set in the schema is `false`.
This config property is accessed in `util/__init__.py`, with `False` as
the default. Thus, `None` should not be the default.

jdavcs committed:
Fix default assignment for `shed_tool_data_path`

As per schema and the logic in `config/__init__.py`,
`shed_tool_data_path` should default to the value of `tool_data_path`
(which is also a reasonable expectation). Currently, "tool-data" is set
as the default value for both properties; however, if `tool_data_path`
is given a different value (e.g., in kwargs at startup) and
`shed_tool_data_path` is not given a value, `shed_tool_data_path` will
no longer default to `tool_data_path`. Removing the explicit default of
`shed_tool_data_path` fixed this issue.

jdavcs committed:
Simplify visualization_plugins_directory assignment

`self.visualization_plugins_directory` is never `None` (default value
"config/plugins/visualizations" will be loaded), so this:
`if ie_dirs and not self.visualization_plugins_directory:`
is never the case.

Also, remove redundant 'defaults to...' comment from option description
in the schema.

jdavcs committed:
Remove redundant assignments

These defauls are already loaded.

jdavcs committed:
Load `cluster_files_directory` default from schema

Note: the call to self.resolve_path() is most likely redundant: data_dir
is already resolved. However, I'll check this some more before making
this change.

jdavcs committed:
Simplify `self.blacklist_file` assignment

`self.blacklist_file` is never `None` (default value
"config/disposable_email_blacklist.conf" will be loaded), so there's no
need to check if it's `None`.

CONSIDERATIONS/TODO:
1. Maybe it makes sense to remove the default value from the schema and,
instead, mention it as a suggested value in the description. Otherwise,
the file will be ALWAYS opened at startup, even when it's empty. IO is
expensive.

2. Error handling IO is questionable: shouldn't we raise a
ConfigurationError here?

jdavcs commented:
Rebased.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -339,40 +333,37 @@ def _process_config(self, kwargs):
-        self.instance_resource_url = kwargs.get('instance_resource_url', None)
+
         #  Get the disposable email domains blacklist file and its contents
-        self.blacklist_location = kwargs.get('blacklist_file', None)
```
The `config/disposable_email_blacklist.conf.sample` sample file is not usable as it is, so the default for `blacklist_file` is actually `None`.
With these changes, `ERROR:galaxy.config:CONFIGURATION ERROR: Can't open supplied blacklist file from path: /home/circleci/repo/config/disposable_email_blacklist.conf` will always be logged at startup, see e.g. https://circleci.com/gh/galaxyproject/galaxy/38709

I see 2 possible solutions:
- change the default in `lib/galaxy/webapps/galaxy/config_schema.yml` to null
- comment out the present content of `config/disposable_email_blacklist.conf.sample` and update the code below to skip lines starting with `#`

jdavcs replied:
I'll remove the default from config_schema.yaml - that will resolve to None automatically. I think it's simple and clean. (+ an edit in `config/__init__.py` to prevent `os.path.join` on None)
Thanks!

jdavcs commented:
Rebased to accommodate new commits to #8648 (addressing reviewer comment)

jdavcs committed:
Remove default value from `blacklist_file`

Addresses reviewer comment:
https://github.com/galaxyproject/galaxy/pull/8651#discussion_r326738014

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -823,15 +860,15 @@ mapping:
           URL of the support resource for the galaxy instance.  Used in activation
           emails.
 
+          Sample default 'https://galaxyproject.org/'
```
```suggestion
          Example value 'https://galaxyproject.org/'
```

jdavcs replied:
Sounds good, and I assume you meant this for all instances of 'Sample default' (correct me if you meant just this one)

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -823,15 +860,15 @@ mapping:
+
       blacklist_file:
         type: str
-        default: config/disposable_email_blacklist.conf
```
I'd re-add this as "Example value" below.

jdavcs replied:
Done

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -454,6 +471,9 @@ mapping:
           container resolver isn't enabled, and will install on demand unless
           involucro_auto_init is set to false.
 
+          Sample default 'database/dependencies/involucro' where 'database/dependencies' is the
```
```suggestion
          Sample default '<tool_dependency_dir>/involucro'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -454,6 +471,9 @@ mapping:
           involucro_auto_init is set to false.
 
+          Sample default 'database/dependencies/involucro' where 'database/dependencies' is the
+          default value of the 'tool_dependency_dir' option.
```
```suggestion
```

jdavcs replied:
Done. (as per previous comment/edit)

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,18 +571,22 @@ mapping:
-          File containing old-style genome builds
+          File containing old-style genome builds.
+
+          Sample default 'tool-data/shared/ucsc/builds.txt' where 'tool-data' is the default value
```
```suggestion
          Sample default '<tool_data_path>/shared/ucsc/builds.txt'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,18 +571,22 @@ mapping:
+          File containing old-style genome builds.
+
+          Sample default 'tool-data/shared/ucsc/builds.txt' where 'tool-data' is the default value
+          of the 'tool_data_path' option.
```
```suggestion
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,18 +571,22 @@ mapping:
         desc: |
           Directory where chrom len files are kept, currently mainly used by trackster
 
+          Sample default 'tool-data/shared/ucsc/chrom' where 'tool-data' is the default value of the
```
```suggestion
          Sample default '<tool_data_path>/shared/ucsc/chrom'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,18 +571,22 @@ mapping:
           Directory where chrom len files are kept, currently mainly used by trackster
 
+          Sample default 'tool-data/shared/ucsc/chrom' where 'tool-data' is the default value of the
+          'tool_data_path' option.
```
```suggestion
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -610,8 +632,7 @@ mapping:
           so proceed with caution. The path is relative to the Galaxy root dir.  To use
-          an absolute path begin the path with '/'.  This is a comma
-          separated list.
+          an absolute path begin the path with '/'.  This is a comma separated list.
```
```suggestion
          an absolute path begin the path with '/'.  This is a comma-separated list.
```

jdavcs replied:
Yes. Compound adjective before a noun. Done.

jdavcs committed:
Wording: Sample default > Example value

In response to reviewer suggestion

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -464,11 +484,12 @@ mapping:
 
       mulled_channels:
         type: str
-        default: 'conda-forge,bioconda'
```
I think this is actually the default value:

https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L453

If you restore this, you can simplify `lib/galaxy/config/__init__.py` I think.

jdavcs replied:
It's one of the edits I wasn't sure about. If we do this (I agree, it makes perfect sense), we should also remove this: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/deps/mulled/mulled_build.py#L42
and 8 occurrences in 3 other files. If you agree, I'll do this as a separate PR.

jdavcs replied:
and instead of those occurrences, simply use the value loaded at startup (i.e., the schema default unless overwritten) that is stored in app.config

nsoranzo replied:
I don't think you can remove it from `lib/galaxy/tool_util/deps/mulled/mulled_build.py`, that's part of galaxy-tool-util package which is used outside of Galaxy (and so doesn't use Galaxy config files).

jdavcs replied:
Oh, I see. So those channels would be specified in 2 places.. On the other hand that gets rid of an import; and that value, I suppose, would very rarely need to be modified, if ever. Yes, I guess that's a reasonable tradeoff. Changing.

jdavcs committed:
Mics edits to schema in response to code review

jdavcs committed:
Puncuation edit in schema: comma-separated

In response to reviewer comment (and grammar rules, of course)

jdavcs committed:
Use schema default for mulled_channels

This simplifies `config/__init__.py` as the value will be assigned by
default.

See this discussion for context:
https://github.com/galaxyproject/galaxy/pull/8651#pullrequestreview-291844458

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -274,13 +290,14 @@ mapping:
 
       conda_prefix:
         type: str
-        default: <tool_dependency_dir>/_conda
```
I was thinking, should this instead be set to `default: _conda`, to be interpreted w.r.t. to `self.tool_dependency_dir` ?

Same goes for `tool_dependency_cache_dir` and `involucro_path`

jdavcs replied:
Yes. And I believe there are more like that. I haven't touched them for the sake of keeping this PR small enough. I'd rather make such additional changes in a follow-up. I'll make this edit, but I'd rather wait with similar edits until this one gets merged.

jdavcs replied:
Correction. I do prefer to leave this until these are merged. `conda_prefix` is not used in `config/__Init__.py`, but it is used (and resolved) in `tool_util/deps/resolvers/conda.py` and `tool_util/deps/conda_util.py`. If/when we set this default, we should also revise how that value is resolved in those two files. I believe this should be left for a different PR. I'm adding it to my TODO list.

jdavcs replied:
Marking as resolved for now, to cleanup this PR. Will make this change in a follow-up. (if you'd rather leave this as not resolved for now, that's fine - please reopen.)

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,18 +569,20 @@ mapping:
 
       builds_file_path:
         type: str
-        default: tool-data/shared/ucsc/builds.txt
```
Similarly, this could be `default: shared/ucsc/builds.txt`, to be interpreted w.r.t. `self.tool_data_path`

Same goes for `len_file_path`

jdavcs replied:
Same as above.

jdavcs replied:
These two done.

jdavcs committed:
Edit schema defaults in response to code review

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -551,17 +569,18 @@ mapping:
 
       len_file_path:
         type: str
-        default: tool-data/shared/ucsc/chrom
```
Add:
```yaml
        default: shared/ucsc/chrom
```

jdavcs replied:
well, that was stupid. Thanks. Done, pushing.

jdavcs committed:
Correct error in previous commit: add default

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -258,36 +250,37 @@ def _process_config(self, kwargs):
-        self.tool_path = resolve_path(kwargs.get("tool_path", "tools"), self.root)
-        self.tool_data_path = resolve_path(kwargs.get("tool_data_path", "tool-data"), os.getcwd())
+        self.tool_path = os.path.join(self.root, self.tool_path)
+        self.tool_data_path = os.path.join(os.getcwd(), self.tool_data_path)
```
```suggestion
        self.tool_data_path = os.path.join(self.root, self.tool_data_path)
```

jdavcs replied:
Done.

jdavcs committed:
Edit in response to review: os.getcwd() > self.root

jdavcs committed:
Fix error in past rebase

Correct parentheses nesting (propogated from prev. rebase)

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -258,36 +250,37 @@ def _process_config(self, kwargs):
-        self.builds_file_path = resolve_path(kwargs.get("builds_file_path", os.path.join(self.tool_data_path, 'shared', 'ucsc', 'builds.txt')), self.root)
-        self.len_file_path = resolve_path(kwargs.get("len_file_path", os.path.join(self.tool_data_path, 'shared', 'ucsc', 'chrom')), self.root)
+            self.tool_data_path = os.path.join(self.data_dir, "tool-data")
+        if self.builds_file_path is None:
```
```suggestion
        self.builds_file_path = os.path.join(self.tool_data_path, self.builds_file_path)
```

jdavcs replied:
I'll commit this suggestion, but here's my reasoning for testing for None.
Currently, this cannot be None. However, if it is, `os.path.join` will, of course, complain. And here's how it could become not None: although a default value is currently assigned, we should be able to override that with a null (see https://github.com/galaxyproject/galaxy/pull/8648#issuecomment-532300644 + preseeding comment by @jmchilton).  I will implement this in a follow-up PR. Also, once we iron out #8651, most paths requiring resolving will be handled differently. The old version had the default hardcoded in this file, so it was safe. I'm OK not checking for None (for now), as this will be rewritten very soon. (Please correct me if this reasoning is incorrect.)

nsoranzo replied:
I don't get what's the use case for allowing these to be `None`, to be honest. I'd not go down that route if it's not necessary.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -258,36 +250,37 @@ def _process_config(self, kwargs):
-        self.len_file_path = resolve_path(kwargs.get("len_file_path", os.path.join(self.tool_data_path, 'shared', 'ucsc', 'chrom')), self.root)
+            self.tool_data_path = os.path.join(self.data_dir, "tool-data")
+        if self.builds_file_path is None:
+            self.builds_file_path = os.path.join(self.root, self.tool_data_path, 'shared', 'ucsc', 'builds.txt')
```
```suggestion
```

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -258,36 +250,37 @@ def _process_config(self, kwargs):
+            self.tool_data_path = os.path.join(self.data_dir, "tool-data")
+        if self.builds_file_path is None:
+            self.builds_file_path = os.path.join(self.root, self.tool_data_path, 'shared', 'ucsc', 'builds.txt')
+        if self.len_file_path is None:
```
```suggestion
        self.len_file_path = os.path.join(self.tool_data_path, self.len_file_path)
```

jdavcs replied:
Same as previous edit.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -258,36 +250,37 @@ def _process_config(self, kwargs):
+        if self.builds_file_path is None:
+            self.builds_file_path = os.path.join(self.root, self.tool_data_path, 'shared', 'ucsc', 'builds.txt')
+        if self.len_file_path is None:
+            self.len_file_path = os.path.join(self.root, self.tool_data_path, 'shared', 'ucsc', 'chrom')
```
```suggestion
```

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -435,22 +422,19 @@ def _process_config(self, kwargs):
-            self.mulled_channels = DEFAULT_CHANNELS
+        # tool_dependency_dir can be "none" (in old configs). If so, set it to schema default
+        if self.tool_dependency_dir and self.tool_dependency_dir.lower() == 'none':
+            self.tool_dependency_dir = self.appschema['tool_dependency_dir'].get('default')
```
According to `tool_dependency_dir` documentation in `lib/galaxy/webapps/galaxy/config_schema.yml` , it seems `None` and `none` can be used to disable tool dependency resolution, see also https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/deps/__init__.py#L51-L52

Ping @jmchilton 

```suggestion
            self.tool_dependency_dir = None
```

jdavcs replied:
Yes, this makes sense. Not sure what I was thinking.

jdavcs replied:
It looks like this change broke some integration tests: in the old version tool_dependency_dir was never None, but now it can be - and so `os.path.join` got a NoneType as a result. I'll fix this today.

nsoranzo replied:
Also, after this line, should `self.tool_dependency_dir` be resolved w.r.t. to `self.data_dir` if it is not None?
At the moment it's only used for the involucro subdir, but maybe better to have it fixed if it is more used in the future?

jdavcs replied:
Very possibly. But I'd rather not do it here. I don't understand its use well enough, and it is accessed + used for path resolving in multiple files. It was not resolved here prior to these changes, so I'd rather leave it as is for now: I am simply not sure I wouldn't be breaking something. Also, I'd rather not introduce new changes in the same PR, if possible.

nsoranzo commented:
It's getting difficult for me as reviewer (and I suppose also for you due to the rebases) to keep track of the stacked PRs.
Would it make sense to close the old ones and concentrate on one?

natefoo commented:
I merged the other PRs, so it should now be possible to rebase this on dev and only retain the commits that are unique.

jdavcs commented:
> I merged the other PRs, so it should now be possible to rebase this on dev and only retain the commits that are unique.

Thank you! And I'm addressing the comments for this PR at the moment.
Rebasing now....

jdavcs committed:
Do not check builds_file_path for None

see discussion in code review

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

jdavcs committed:
Update lib/galaxy/config/__init__.py

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

jdavcs committed:
Do not check len_file_path for None

See discussion in code review

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

jdavcs committed:
Update lib/galaxy/config/__init__.py

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

jdavcs committed:
If tool_dependency_dir is 'none', do not use default

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -1637,7 +1634,7 @@ mapping:
           {pid}
 
-          Sample default 'heartbeat_{server_name}.log'
+          Example value 'heartbeat_{server_name}.log'
```
```suggestion
          Sample default 'heartbeat_{server_name}.log'
```

jdavcs replied:
I see, that makes sense. I believe some of them may be handled with `path_resolves_to` once that is ready, but for now I'll change them all.

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -52,7 +52,7 @@ mapping:
           config file documentation.
 
-          Sample default 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
+          Example value 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
```
```suggestion
          Sample default 'sqlite:///<data_dir>/universe.sqlite?isolation_level=IMMEDIATE'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -52,7 +52,7 @@ mapping:
 
-          Sample default 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
+          Example value 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
           where './database' is the resolved default value of the 'data_dir' option).
```
```suggestion
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -137,7 +137,7 @@ mapping:
           but prefixed with install_ are also available).
 
-          Sample default 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
+          Example value 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
```
```suggestion
          Defaults to the value of the 'database_connection' option.
```

jdavcs replied:
Yes, that's much better. Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -295,7 +295,7 @@ mapping:
           installed.
 
-          Sample default '<tool_dependency_dir>/_conda'
+          Example value '<tool_dependency_dir>/_conda'
```
```suggestion
          Sample default '<tool_dependency_dir>/_conda'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -376,7 +376,7 @@ mapping:
           of the tool dependency directory.
 
-          Sample default '<tool_dependency_dir>/_cache'
+          Example value '<tool_dependency_dir>/_cache'
```
```suggestion
          Sample default '<tool_dependency_dir>/_cache'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -470,8 +470,7 @@ mapping:
 
-          Sample default 'database/dependencies/involucro' where 'database/dependencies' is the
-          default value of the 'tool_dependency_dir' option.
+          Example value '<tool_dependency_dir>/involucro'
```
```suggestion
          Sample default '<tool_dependency_dir>/involucro'
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -571,21 +568,19 @@ mapping:
-          Sample default 'tool-data/shared/ucsc/builds.txt' where 'tool-data' is the default value
-          of the 'tool_data_path' option.
+          File containing old-style genome builds. Value will be resolved to
+          '<tool_data_path>/shared/ucsc/builds.txt'.
```
```suggestion
          with respect to <tool_data_path>.
```

jdavcs replied:
Done. There may be other instances of "resolved to" versus "resolved with respect to". I agree the latter form is more correct; so that would be another minor todo item for later.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -571,21 +568,19 @@ mapping:
-          Sample default 'tool-data/shared/ucsc/chrom' where 'tool-data' is the default value of the
-          'tool_data_path' option.
+          Directory where chrom len files are kept, currently mainly used by trackster. Value will
+          be resolved to '<tool_data_path>/shared/ucsc/chrom'.
```
```suggestion
          be resolved with respect to <tool_data_path>.
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -672,12 +665,15 @@ mapping:
           If using a cluster, Galaxy will write job scripts and stdout/stderr to this
           directory.
 
+          Default value will be resolved to 'database/pbs' where 'database' is the
```
```suggestion
          Value will be resolved with respect to <data_dir>.
```

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -672,12 +665,15 @@ mapping:
           directory.
 
+          Default value will be resolved to 'database/pbs' where 'database' is the
+          default value of the 'data_dir' option).
```
```suggestion
```

jdavcs replied:
and done.

nsoranzo reviewed (commented):
I think that "Sample default" is better for values that cannot be set in `config_schema.yml` because they are not static, but there is actually a default value.
On the other hard, "Example value" is better when there is no default value (static or calculated).

jdavcs committed:
Fix involucro path resolution

If `tool_dependency_dir` is None, use schema default for path resolving

jdavcs committed:
Edit wording in schema as per code review

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -53,7 +53,6 @@ mapping:
           config file documentation.
 
           Sample default 'sqlite:///./database/universe.sqlite?isolation_level=IMMEDIATE'
```
I think you missed this one:

```suggestion
          Sample default 'sqlite:///<data_dir>/universe.sqlite?isolation_level=IMMEDIATE'
```

jdavcs replied:
Fixing it now.

jdavcs committed:
Update lib/galaxy/webapps/galaxy/config_schema.yml

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -291,16 +288,16 @@ def _process_config(self, kwargs):
-        self.tour_config_dir = os.path.join(self.root, kwargs.get("tour_config_dir", "config/plugins/tours"))
-        self.webhooks_dirs = os.path.join(self.root, kwargs.get("webhooks_dir", "config/plugins/webhooks"))
+        self.tour_config_dir = os.path.join(self.root, self.tour_config_dir)
+        self.webhooks_dirs = os.path.join(self.root, self.webhooks_dir)
```
I just noticed that:
- `webhooks_dir` can be a comma-separated list of directories, so resolving a list w.r.t `self.root` would affect only the first element of the list.
- the assigned attribute `self.webhooks_dirs` has a final `s`, while the config property doesn't

Given that each element of the list is resolved w.r.t. `self.root` any way in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/util/__init__.py#L1584 , I'd suggest to remove this line:

```suggestion
```

and apply this one-line patch:
```patch
diff --git a/lib/galaxy/app.py b/lib/galaxy/app.py
index bd1de55caa..726c3380f3 100644
--- a/lib/galaxy/app.py
+++ b/lib/galaxy/app.py
@@ -155,7 +155,7 @@ class UniverseApplication(config.ConfiguresGalaxyMixin):
         # Tours registry
         self.tour_registry = ToursRegistry(self.config.tour_config_dir)
         # Webhooks registry
-        self.webhooks_registry = WebhooksRegistry(self.config.webhooks_dirs)
+        self.webhooks_registry = WebhooksRegistry(self.config.webhooks_dir)
         # Load security policy.
         self.security_agent = self.model.security_agent
         self.host_security_agent = galaxy.model.security.HostAgent(
```

jdavcs replied:
Would you mind if I do this as a separate PR? I can do it right now - will take a couple of minutes.. 
The reason I'm asking is (1) it's not fixing something this PR modified (so the potential list of such edits is very long); and (2) it's ~20 minutes away from passing the last test after which it can be merged - then I can finally rebase 8689. If I make this change, it'll launch all the tests all over again - and that'll take hours.

nsoranzo replied:
Sure, you can address all 3 remaining comments in #8689 or a separate PR.

jdavcs replied:
Thank you! Will do.

jdavcs replied:
I believe the same applies to `tour_registry` (except the trailing 's'). Same thing: comma-separated list, resolved in the same way. I'm changing both.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -354,17 +351,17 @@ def _process_config(self, kwargs):
+                with open(self.blacklist_file) as f:
+                    self.blacklist_content = [line.rstrip() for line in f]
             except IOError:
                 log.error("CONFIGURATION ERROR: Can't open supplied blacklist file from path: " + str(self.blacklist_file))
```
```suggestion
                log.error("CONFIGURATION ERROR: Can't open supplied blacklist file from path: %s", self.blacklist_file)
```
